File: internal/maintenance/cat_filehistory.go
package maintenance

import (
	"context"
	"path/filepath"
)

func init() { registerMatcher("file-history", 30, scanFileHistory) }

// scanFileHistory surfaces reclaimable storage under <root>/file-history — the
// CLI's per-file edit-undo snapshot store, one UUID dir per tracked file with
// no built-in retention. Candidates are split into two groups:
//   - "empty": a UUID dir holding zero snapshot files (dead weight, no undo
//     value at all).
//   - "stale": a UUID dir whose newest snapshot has aged past the cutoff
//     (default 30d) — undo history that old is unlikely to ever be restored.
//
// Age comes from the newest descendant mtime (subtreeStats), not the dir's
// own mtime, so an unrelated metadata touch can't mask a genuinely stale dir.
func scanFileHistory(ctx context.Context, spec CategorySpec) ([]Candidate, error) {
	dir := filepath.Join(spec.Root, "file-history")
	entries, ok, err := openDirNoSymlink(dir)
	if err != nil || !ok {
		return []Candidate{}, err
	}

	out := []Candidate{}
	for _, e := range entries {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		if !e.IsDir() {
			continue // file-history/ is per-UUID dirs; stray files are not ours to judge
		}
		uuidDir := filepath.Join(dir, e.Name())
		bytes, files, newest, err := subtreeStats(ctx, uuidDir)
		if err != nil {
			return nil, err
		}

		if files == 0 {
			out = append(out, Candidate{
				Path: uuidDir, Bytes: bytes, Files: files, ModTime: newest,
				Reason: "empty history dir", Group: "empty",
				Meta: map[string]string{"uuid": e.Name()},
			})
			continue
		}
		if !olderThan(newest, spec) {
			continue
		}
		out = append(out, Candidate{
			Path: uuidDir, Bytes: bytes, Files: files, ModTime: newest,
			Reason: "no edits in 30+ days", Group: "stale",
			Meta: map[string]string{"uuid": e.Name()},
		})
	}
	return out, nil
}

File: src-tauri/src/maintenance/cat_filehistory.rs
//! Ported from `internal/maintenance/cat_filehistory.go` (W13). Surfaces
//! reclaimable storage under `<root>/file-history` (per-UUID edit-undo snapshot
//! dirs, no built-in retention). Split into `empty` (zero snapshot files) and
//! `stale` (newest snapshot aged past the cutoff, default 30d). Age comes from
//! the newest descendant mtime, not the dir's own.

use std::collections::BTreeMap;
use std::path::Path;

use super::category::{older_than, open_dir_no_symlink, subtree_stats};
use super::types::{Candidate, CategorySpec};

pub fn scan_file_history(spec: &CategorySpec) -> Result<Vec<Candidate>, String> {
    let dir = Path::new(&spec.root).join("file-history");
    let (entries, ok) = open_dir_no_symlink(&dir)?;
    if !ok {
        return Ok(Vec::new());
    }

    let mut out: Vec<Candidate> = Vec::new();
    for e in &entries {
        if !e.file_type().map(|t| t.is_dir()).unwrap_or(false) {
            continue; // per-UUID dirs; stray files are not ours to judge
        }
        let name = e.file_name().to_string_lossy().into_owned();
        let uuid_dir = dir.join(&name);
        let (bytes, files, newest) = subtree_stats(&uuid_dir);

        let mut meta = BTreeMap::new();
        meta.insert("uuid".to_string(), name.clone());

        if files == 0 {
            out.push(Candidate {
                path: uuid_dir.to_string_lossy().into_owned(),
                bytes,
                files,
                mod_time: newest,
                reason: "empty history dir".to_string(),
                group: "empty".to_string(),
                meta,
            });
            continue;
        }
        if !older_than(newest, spec) {
            continue;
        }
        out.push(Candidate {
            path: uuid_dir.to_string_lossy().into_owned(),
            bytes,
            files,
            mod_time: newest,
            reason: "no edits in 30+ days".to_string(),
            group: "stale".to_string(),
            meta,
        });
    }
    Ok(out)
}

#[cfg(test)]
#[path = "cat_filehistory_tests.rs"]
mod cat_filehistory_tests;
